### PR TITLE
Move cell movement into a descriptor

### PR DESCRIPTION
**Pull request by quaquel**
This PR encapsulates the addition to and removal from a cell of an agent in a `HasCell` mixin. It updates CellAgent to use this mixin, and adds an override of `agent.remove` to ensure agents remove themselves from the cell they are occupying when they are removed from the model. The main implication of this PR is that `CellAgent.move_to` is now redundant. I have thus removed it. In #2292, we might bring it back but with more powerful behavior.

### outdated based on discussion below
This PR is for discussion in relation to #2292. It shows how it might be possible to use composition for cell-related movement by encapsulating all the movement behavior into a descriptor. So instead of having a special `Agent.move_to` method, you can do `Agent.cell = new_cell`. You can also use whatever name you want instead of `cell`.

An additional benefit that I had not realized at first is that it cleans up the API a bit more as well. A user does not have to interact with the Cell class directly. She only needs to assign the Cell instance to whatever attribute is used for the descriptor.

### CODE REVIEW

**Comment by EwoutH** [comment]
I think I like this, it makes it much more explicit what's happening. It's like modifying `pos` directly, but for the new spaces.

One thing I'm curious about if `cell` has now a special status. How does the space know what Agent variable to lookup / modify? What is there are multiple spaces?

I will review in more detail tomorrow!

**Comment by Corvince** [comment]
Ha! Given enough time I think @quaquel will turn every mesa behaviour into a descriptor 😂

No, I really like this and this somehow put me over the edge to finally try to fully grok descriptors. 

Let me write in more detail when I have more time, but I think using just a property with a setter might be better

**Comment by quaquel** [comment]
> One thing I'm curious about if cell has now a special status. How does the space know what Agent variable to lookup / modify? What is there are multiple spaces?

A space does not need to know anything about Agents. A space has cells. A cell can have agents. Agents and cells need to be aware of each other. A space is primarily a container for cells, nothing more. Multiple spaces should not be a problem. You just again another cell to the attribute and your done. 


> No, I really like this and this somehow put me over the edge to finally try to fully grok descriptors.

The shortest summary is "reusable properties". 

> Let me write in more detail when I have more time, but I think using just a property with a setter might be better

I think you need a getter because the agent needs to know its cell for e.g., a random walk, but curious to see what more you have to say in detail. 


Given the fact that both @Corvince and @EwoutH like this idea, I'll try to update the tests asap to make sure this is fully covered again.

**Comment by EwoutH** [comment]
(I just looked at it from my phone, still need to do a thorough review)

**Comment by Corvince** [comment]
> I think you need a getter because the agent needs to know its cell for e.g., a random walk, but curious to see what more you have to say in detail.

My main objection would be that it might make some things easier if we have a clear "cell" definition. That is with a reusable descriptor we have the advantage of allowing several "cell" descriptor from several spaces. In practice I think this isn't very common. 

In the case of a "normal" property we could have the following:

```python
class HasCell:
    _cell: Cell | None = None

    @property
    def cell(self):
        return self._cell

    @cell.setter
    def cell(self, cell: Cell):
        ... # Your logic from CellDescriptor.__set__
```

And then we can have separate movement mixin
```python
class CellMovement:
    def move_to(self, cell: Cell):
         self.cell = cell

    ... # More movement functions
```

And finally the CellAgent:

```python
class CellAgent(Agent, HasCell, CellMovement):
    ....
```

In this case its unambigious and easy. We know we have to assign to "cell". If the cell descriptor can be assigned to any name, how would we know the right name?

Further advantages are that one could theoretically overwrite the `cell.setter` function. Also I think more people know properties in comparison to descriptors (although the former is just a higher level abstraction).

**Comment by quaquel** [comment]
> In this case its unambigious and easy. We know we have to assign to "cell". If the cell descriptor can be assigned to any name, how would we know the right name?

Not sure I fully understand your point, but a descriptor has the [`__set_name__` magic method](https://docs.python.org/3/howto/descriptor.html#customized-names). This magic method is called during class creation with the owning class and name of the attribute to which the descriptor is assigned. So, with a descriptor, you can always know the name of the attribute. 

Also, the `__set__` method in the descriptor would make a `move_to` method completely redundant.

I like your name `HasCell` as a possible name for the descriptor. It is probably better than my current uninspiring `CellDescriptor`

**Comment by Corvince** [comment]
Sorry if I wasn't clear. The movement mixin was just an example (although I would like to have the clearer move_to API, even if redundant. But this is irrelevant for my point here). The point is that for any mixin or rather anything wanting to operate on the cell property it needs to know the name. Of course the user knows the name, but our framework doesn't. So in the example move_to method the line `self.cell = cell` would not work if the name is not cell, but something else.

**Comment by quaquel** [comment]
> Sorry if I wasn't clear. The movement mixin was just an example (although I would like to have the clearer move_to API, even if redundant. But this is irrelevant for my point here). The point is that for any mixin or rather anything wanting to operate on the cell property it needs to know the name. Of course the user knows the name, but our framework doesn't. So in the example move_to method the line self.cell = cell would not work if the name is not cell, but something else.

Ok, that makes sense. If we want to add any additional functionality like the MoveTo example, then yes we need to lock down the name of the attribute to which the current cell is assigned.

**Comment by quaquel** [comment]
A quick follow-up to @Corvince point about the naming of the attribute. There is, in fact, already a case like this. It is up to the user to actively remove an agent from the cell before calling `Agent.remove`. Moving this behavior into `CellAgent.remove` would make sense, but this method needs to know the attribute to which `Cell` is assigned. So the descriptor idea, which leaves the name of the attribute to the user, is less than ideal. 

This leaves the other idea: instead of having a `move_to` method solve everything via attribute assignment: `self.cell = self.cell.neighborhood.select_random_cell()`, which can be done with properties as shown by the HasCell mixin shown by @Corvince . I do like the clean API that results from solving cell movement this way and slightly prefer it over the current move_to method. 

Of course this still leaves room for more sophisticated movement methods that for example use a heading or something.

**Comment by quaquel** [comment]
> One thing I would find interesting is how this would work with multiple grids.

1. I think this question is only tangentially related to the point of this PR/discussion.
2. Do we want to support that out of the box with MESA? I personally don't think so.

**Comment by EwoutH** [comment]
Just making sure we don’t box ourselves in / prohibit a use case unintentionally.

**Comment by Corvince** [comment]
If we would want to provide strong support, it would be relevant for this PR. But I also think we must not provide that out of the box. 

For me the space should be uniquely defined. A single object can't exist in two places at once. Everything else is a special case/abstraction. And it's probably special enough not to deal with it preemptively. But of course we must observe this and if we see common patterns we can abstract over that.

And by the way the cell space is super flexible with its connections. It's actually straight forward to have connections to the local spatial neighborhood, but also connect cells to their "network neighbors". Given that we now have "named connections" it's even easy to differentiate between both types.

**Comment by quaquel** [comment]
> I very much agree with this as the underlying low-level implementation. It's also very much the intuitive way to assign a new position. So from a model or space point of view that's the way to do it.
> 
> But from an agent perspective I would also like to have a move_to function that semantically explicitly moves an agent from one location/cell to another. Even if that's just an assignment under the hood. But given the different perspectives I don't think this violates the "one obvious way to do" recommendation.


Ok, I'll update this PR accordingly. Not yet sure about keeping `move_to` but I am fine with keeping it for now.

**Comment by Corvince** [comment]
Just to clarify, I would not make it part of HasCell, but in a separate movement mixin, probably with other movement options, thus also out of scope of this PR

**Comment by Corvince** [comment]
Yes, I see this as a perfect base case for the ideas discussed in #2292. After this is merged I would update #2292 to incorporate the  named coordinates and create a "Movement" mixin class. Then CellAgent can become just a class inherting from Agent, HasCell and Movement without any MRO related headaches (because of those, only Agent has an __init__ method). I think this will come out nicely, but once implemented we can see how it plays out by adapting some examples.

**Comment by Corvince** [comment]
There are 6 ruff errors otherwise this looks good to go from my side (but a lot of things are based on my ideas so I would like another reviewer to approve)

**Comment by rht** [comment]
Disclaimer: I haven't read #2292. I'm not fully sure why having a `move_to` method makes it not compose-able, but from UX perspective, I prefer for it to be not removed. It's a lot easier to search and understand the codebase if there is an actual `move_to` verb, and having to grep for `self.cell = self.cell.*` for an agent movement (this is a relatively more complex expression regexp vs just a keyword `move_to`), making it as a too Mesa-specific quirk and not as modeling in general.

While @quaquel 's stance on accommodating LLM can be found at https://github.com/projectmesa/mesa/pull/2237#issuecomment-2304032222, I find that making it less likely for LLM to fall into a gotcha / to help it to reason better actually doubles as to help human reasoners as well.

Why can't we have both `move_to` and while having the descriptor under the hood?

**Comment by quaquel** [comment]
> Why can't we have both move_to and while having the descriptor under the hood?

The idea is that `move_to` will be made more powerful via #2292. So yes we'll likely end up with both. The key thing here is that the logic of adding and removing agents from cells is encapsulated in the attribute assignment rather than via a separate method.

**Review by Corvince — APPROVED** [review]

File: benchmarks/Schelling/schelling.py
Comment (rht) [inline benchmarks/Schelling/schelling.py:45]: If we have both, then why is this not `self.move_to`?

Comment (quaquel) [reply benchmarks/Schelling/schelling.py:45]: please read the discussion both here and in #2292.

Comment (quaquel) [reply benchmarks/Schelling/schelling.py:45]: I am sorry for the short answer and the merge, but move_to is added back in via #2292.

Comment (rht) [reply benchmarks/Schelling/schelling.py:45]: I see, then you have my post-merge approval.

**Review by EwoutH — COMMENTED** [review]
Thanks for working on this. This looks really robust, and it looks like it's more robust than previous movement methods.

I left a few thoughts when going through the diff.

File: mesa/experimental/cell_space/cell_agent.py
Comment (quaquel) [reply mesa/experimental/cell_space/cell_agent.py:19]: As per @Corvince's point above, we don't want that. We need to know the name of the attribute so other mixins can rely on it.

File: tests/test_cell_space.py
Comment (quaquel) [reply tests/test_cell_space.py:553]: In this case I explicitly am testing for the hard reference because if that is gone, I know that the object can be garbage collected. 

The syntax in general is indeed valid.

File: mesa/experimental/cell_space/cell_agent.py
Comment (quaquel) [reply mesa/experimental/cell_space/cell_agent.py:36]: Let's pick up that point in #2292

File: mesa/experimental/cell_space/cell_agent.py
Comment (EwoutH) [reply mesa/experimental/cell_space/cell_agent.py:19]: Yeah, sounds about right, I also concluded that when developing my previous proposal.

At some point we have to document that.

File: benchmarks/WolfSheep/wolf_sheep.py
Comment (quaquel) [reply benchmarks/WolfSheep/wolf_sheep.py:118]: Oh I might better have left that change out. Basically, there was a fixme statement in the docstring to go from 2 arguments (countdown value and whether or not being fully grown) to only 1. If countdown is 0, you are fully grown, otherwise you are not.

**Pull request merged**
